fix(ivy): DebugNode.attributes not preserving attribute name casing

crisbeto opened the pull request:
In Ivy the `DebugNode.attributes` is populated directly from the DOM, but the problem is that the browser will lowercase all attribute names. These changes preserve the case by first going through the `TNode.attrs`, populating the map with the case-sensitive names and saving a reference to the lower case name. Afterwards when we're going through the attributes from the DOM, we can check whether we've mapped the attribute by its case-sensitive name already.

This PR resolves FW-1358.

crisbeto commented:
Merge assistance: Codefresh seems to be flaky again.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -276,13 +277,64 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
+        if (attrName === AttributeMarker.NamespaceURI) {
+          // Skip namespaced attributes since the browser should preserve their casing.
+          i += 4;
+          continue;
```
Do we need to `continue` here? I think we can just brake since the first marker indicates that there are no "regular" attributes here. See `setUpAttributes` in the `attrs_utils.ts`

crisbeto replied:
I wasn't sure whether we're guaranteed that the ones with markers would be all the way at the end, but that appears to be the case. I've removed all the special marker logic in favor of breaking out as soon as we run into a marker.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -276,13 +277,64 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
+      while (i < tNodeAttrs.length) {
+        const attrName = tNodeAttrs[i];
+        if (attrName === AttributeMarker.NamespaceURI) {
+          // Skip namespaced attributes since the browser should preserve their casing.
```
Could you please add a test for this case?

crisbeto replied:
I've added one that namespaced attributes are picked up.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -276,13 +277,64 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
+          // Skip namespaced attributes since the browser should preserve their casing.
+          i += 4;
+          continue;
+        } else if (attrName === AttributeMarker.ProjectAs) {
```
I'm not sure what VE does for this `ngProjectAs` attribute - could you please add a test?

crisbeto replied:
I don't think it does. Since I've changed the logic to break out of the loop, this shouldn't matter anymore.

pkozlowski-opensource commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -276,13 +277,64 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
+      }
+    }
+
+    const eAttrs = element.attributes;
```
So, now I start to wonder what VE does here, really. For example, is it going to return attributes that are set by DOM manipulation? Reflected as a side-effect of prop binding?

We should add a test for those cases.

crisbeto replied:
I think I worded the comment above poorly. VE won't pick up something like `el.setAttribute('foo', 'bar')` either, but it will pick up attributes set through the `Renderer2`. I've added a test case for those.

pkozlowski-opensource reviewed (changes requested):
Left some in-line comments, it mostly needs more tests.

crisbeto committed:
fix(ivy): DebugNode.attributes not preserving attribute name casing

In Ivy the `DebugNode.attributes` is populated directly from the DOM, but the problem is that the browser will lowercase all attribute names. These changes preserve the case by first going through the `TNode.attrs`, populating the map with the case-sensitive names and saving a reference to the lower case name. Afterwards when we're going through the attributes from the DOM, we can check whether we've mapped the attribute by its case-sensitive name already.

This PR resolves FW-1358.

crisbeto reviewed (commented):
Thank you for the feedback @pkozlowski-opensource, it should all be addressed now. Can you take another look?

pkozlowski-opensource reviewed (approved):
LGTM

AndrewKushnir commented:
- [VE presubmit](http://test/OCL:251729095:BASE:251729113:1559772905577:af21d6a6)
- [Ivy presubmit](https://test.corp.google.com/ui#id=OCL:251729371:BASE:251729666:1559773071348:79a7baf1)

AndrewKushnir commented:
Ivy presubmit looks normal, VE presubmit is successful, adding "merge" label.
